dbms/src/IO/Compression/CompressionCodecDeltaPFor.cpp

```diff
@@ -0,0 +1,380 @@
+// limitations under the License.
+
+#include <Common/BitpackingPrimitives.h>
+#include <Common/Exception.h>
+#include <DataTypes/IDataType.h>
+#include <IO/Compression/CompressionCodecDeltaPFor.h>
+#include <IO/Compression/CompressionInfo.h>
+#include <common/likely.h>
+#include <common/unaligned.h>
+
+#if defined(__AVX2__)
+#include <immintrin.h>
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+} // namespace ErrorCodes
+
+CompressionCodecDeltaPFor::CompressionCodecDeltaPFor(UInt8 bytes_size_)
+    : bytes_size(bytes_size_)
+{}
+
+UInt8 CompressionCodecDeltaPFor::getMethodByte() const
+{
+    return static_cast<UInt8>(CompressionMethodByte::DeltaPFor);
+}
+
+UInt32 CompressionCodecDeltaPFor::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    // 1 byte for bytes_size, x bytes for frame of reference, 1 byte for width.
+    return 1 + bytes_size + sizeof(UInt8)
+        + BitpackingPrimitives::getRequiredSize(uncompressed_size / bytes_size, bytes_size * 8);
+}
+
+namespace
+{
+template <typename T>
+UInt32 compressData(const char * source, UInt32 source_size, char * dest)
+{
+    const auto count = source_size / sizeof(T);
+    using ST = typename std::make_signed<T>::type;
+    std::vector<ST> deltas;
+    deltas.reserve(count);
+    T prev_src = 0;
+    const char * const source_end = source + source_size;
+    while (source < source_end)
+    {
+        T curr_src = unalignedLoad<T>(source);
+        deltas.push_back(static_cast<ST>(curr_src - prev_src));
+        prev_src = curr_src;
+        source += sizeof(T);
+    }
+    ST frame_of_reference = *std::min_element(deltas.cbegin(), deltas.cend());
+    // store frame of reference
+    unalignedStore<ST>(dest, frame_of_reference);
+    dest += sizeof(ST);
+    if (frame_of_reference != 0)
+    {
+        for (auto & delta : deltas)
+            delta -= frame_of_reference;
+    }
+    ST max_value = *std::max_element(deltas.cbegin(), deltas.cend());
+    UInt8 width = BitpackingPrimitives::minimumBitWidth(max_value);
+    // store width
+    unalignedStore<UInt8>(dest, width);
+    dest += sizeof(UInt8);
+    // if width == 0, skip bitpacking
+    if (width == 0)
+        return sizeof(ST) + sizeof(UInt8);
+    auto required_size = BitpackingPrimitives::getRequiredSize(count, width);
+    // after applying frame of reference, all deltas are bigger than 0.
+    BitpackingPrimitives::packBuffer(
+        reinterpret_cast<unsigned char *>(dest),
+        reinterpret_cast<const T *>(deltas.data()),
+        count,
+        width);
+    return sizeof(ST) + sizeof(UInt8) + required_size;
+}
+
+template <class T>
+void ApplyFrameOfReference(T * dst, T frame_of_reference, UInt32 count)
+{
+    if (!frame_of_reference)
+        return;
+
+    UInt32 i = 0;
+    UInt32 misaligned_count = count;
+#if defined(__AVX2__)
+    misaligned_count = count % (sizeof(__m256i) / sizeof(T));
```

Comment:
What if size of T is greater than size of __m256i?

Reply:
impossible, the largest is UInt64.

Reply:
static assert that

Reply:
done